File: packages/lib-angular/projects/avatars/src/lib/avatars/avatar-bauhaus.ts
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";

import { Component, Input } from "@angular/core";

import type { AvatarProps } from "../avatar.utils";

import { defaultAvatarProps } from "../avatar.utils";
import { generateColors, SIZE } from "./avatar-bauhaus.utils";

@Component({
  selector: "avatar-bauhaus, AvatarBauhaus",
  template: `
    <svg
      fill="none"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      [attr.viewBox]='"0 0 " + SIZE + " " + SIZE'
      [attr.width]="size"
      [attr.height]="size"
    >
      <ng-container *ngIf="title">
        <title>{{name}}</title>
      </ng-container>

      <mask
        id="mask__bauhaus"
        maskUnits="userSpaceOnUse"
        [attr.x]="0"
        [attr.y]="0"
        [attr.width]="SIZE"
        [attr.height]="SIZE"
      >
        <rect
          fill="#FFFFFF"
          [attr.width]="SIZE"
          [attr.height]="SIZE"
          [attr.rx]="square ? undefined : SIZE * 2"
        ></rect>
      </mask>

      <g mask="url(#mask__bauhaus)">
        <rect
          [attr.width]="SIZE"
          [attr.height]="SIZE"
          [attr.fill]="properties(name, colors)?.[0]?.color"
        ></rect>

        <rect
          [attr.x]="(SIZE - 60) / 2"
          [attr.y]="(SIZE - 20) / 2"
          [attr.width]="SIZE"
          [attr.height]="properties(name, colors)?.[1]?.isSquare ? SIZE : SIZE / 8"
          [attr.fill]="properties(name, colors)?.[1]?.color"
          [attr.transform]="getSquareTransform()"
        ></rect>

        <circle
          [attr.cx]="SIZE / 2"
          [attr.cy]="SIZE / 2"
          [attr.fill]="properties(name, colors)?.[2]?.color"
          [attr.r]="SIZE / 5"
          [attr.transform]="getCircleTransform()"
        ></circle>

        <line
          [attr.x1]="0"
          [attr.y1]="SIZE / 2"
          [attr.x2]="SIZE"
          [attr.y2]="SIZE / 2"
          [attr.strokeWidth]="2"
          [attr.stroke]="properties(name, colors)?.[3]?.color"
          [attr.transform]="getLineTransform()"
        ></line>
      </g>
    </svg>
  `,
})
export class AvatarBauhaus {
  SIZE = SIZE;

  @Input() name: Omit<AvatarProps, 'variant'>['name'] = defaultAvatarProps['name'];
  @Input() colors: Omit<AvatarProps, 'variant'>['colors'] = defaultAvatarProps['colors'];
  @Input() size: Omit<AvatarProps, 'variant'>['size'] = defaultAvatarProps['size'];
  @Input() title: Omit<AvatarProps, 'variant'>['title'] = defaultAvatarProps['title'];
  @Input() square: Omit<AvatarProps, 'variant'>['square'] = defaultAvatarProps['square'];

  properties(name, colors) {
    return generateColors(name, colors);
  }
  getSquareTransform() {
    return (
      "translate(" +
      this.properties(this.name, this.colors)?.[1]?.translateX +
      " " +
      this.properties(this.name, this.colors)?.[1]?.translateY +
      ") rotate(" +
      this.properties(this.name, this.colors)?.[1]?.rotate +
      " " +
      SIZE / 2 +
      " " +
      SIZE / 2 +
      ")"
    );
  }
  getCircleTransform() {
    return (
      "translate(" +
      this.properties(this.name, this.colors)?.[2]?.translateX +
      " " +
      this.properties(this.name, this.colors)?.[2]?.translateY +
      ")"
    );
  }
  getLineTransform() {
    return (
      "translate(" +
      this.properties(this.name, this.colors)?.[3]?.translateX +
      " " +
      this.properties(this.name, this.colors)?.[3]?.translateY +
      ") rotate(" +
      this.properties(this.name, this.colors)?.[3]?.rotate +
      " " +
      SIZE / 2 +
      " " +
      SIZE / 2 +
      ")"
    );
  }
}

@NgModule({
  declarations: [AvatarBauhaus],
  imports: [CommonModule],
  exports: [AvatarBauhaus],
})
export class AvatarBauhausModule {}

File: packages/lib-leptos/src/avatar_bauhaus.rs
use crate::avatar_bauhaus_utils::*;
use crate::avatar_utils::*;
use leptos::*;

#[component]
pub fn AvatarBauhaus(
    cx: Scope,
    name: String,
    colors: Vec<&'static str>,
    #[prop(optional, into)] square: bool,
) -> impl IntoView {
    let num_from_name = hash_code(&name);
    let generated_colors = generate_colors(&name, colors);

    let half_size = SIZE / 2;
    let outer_radius = if square == true { 0.0 } else { (SIZE as f64) * 2.0 };
    let circle_radius = SIZE / 5;
    let rectangle_height= SIZE / 8;
    let rectangle_x = (SIZE - 60) / 2;
    let rectangle_y = (SIZE - 20) / 2;


    view! { cx,
        <svg
            fill="none"
            role="img"
            xmlns="http://www.w3.org/2000/svg"
            viewBox=format!("0 0 {} {}",SIZE, SIZE)
            width=200
            height=200
        >
            <mask
                id=format!("mask__bauhaus_{}", num_from_name)
                maskUnits="userSpaceOnUse"
                x=0
                y=0
                width=SIZE
                height=SIZE
            >
                <rect
                    fill="#FFFFFF"
                    width=SIZE
                    height=SIZE
                    rx=outer_radius
                />
            </mask>
            <g mask=format!("url(#mask__bauhaus_{})", num_from_name)>
                <rect
                    width=SIZE
                    height=SIZE
                    fill=generated_colors[0].color.clone()
                />

                <rect
                    x=rectangle_x
                    y=rectangle_y
                    width=SIZE
                    height=rectangle_height
                    fill=generated_colors[1].color.clone()
                    transform=get_square_transform(generated_colors.clone())
                />

                <circle
                    cx=half_size
                    cy=half_size
                    fill=generated_colors[2].color.clone()
                    r=circle_radius
                    transform=get_circle_transform(generated_colors.clone())
                />

                <line
                    x1=0
                    y1=half_size
                    x2=SIZE
                    y2=half_size
                    strokeWidth=2
                    stroke=generated_colors[3].color.clone()
                    transform=get_line_transform(generated_colors.clone())
                />
            </g>
        </svg>
    }
}
